### PR TITLE
Add fines for users with negative balances

### PR DESCRIPTION
**Pull request by Yoronex**
Like the name suggest, this PR adds fines! It also contains some extra quality of life endpoints to make the life of the BAC treasurer easier.

- Fines are calculated based on the definition in the SudoSOS TOS (1 euro for every 5 euro debt, rounded down (aka 20% rounded down), with a maximum of 5 euros).
- The height of the fine is based on the user's balance on the reference date, but only if they are in debt both on the reference date and now.
- With the `GET /fines/eligible` endpoint, one can get a list of all users having one of the given user types and having a negative balance both now and on the reference date. 
- With the `POST /fines/handout` endpoint, one can hand out fines to all given users. If a user is not eligible for a fine, it will be fined 0 euros.
- Users get an email when they got fined.
- Fines are grouped in FineHandoutEvents (to list all fines that have been handed out in a single session) and in UserFineGroups (all fines that a user collected over time in a single "debt").
- The `User` entity has a reference `currentFines` to the currently applicable `UserFineGroup`. Via a subscriber in the `transfer.ts` entity, this reference will automatically be removed once a user has gone out of debt.
- The amount of fines is returned in the `GET balance` endpoints.
- Single fines can be deleted (removed from history) when the treasurer made a mistake.
- Fines can be waived. A new transfer will be created to "cancel out" the amount fined previously.

Note that the following considerations have been made for the given implementation:
**Reference to currently applicable UserFineGroup in User entity**
Having this reference breaks the philosophy of a normalized database for SudoSOS. However, such a reference is necessary to efficiently and without any extreme complexity query the current amount of fines a user has not yet paid for their current debt. Calculating this using the same method as we currently calculate a user's balance with, is undoable. It is fairly easy to calculate the sum of multiple fines, but to do this correctly we need to know the first fine of the sequence. Again, to know this we will need to know the point in time the user went into debt, which is impossible to do easily, as we then will need to calculate balances over time for all users. With the current method, we can simply do a `LEFT JOIN` on the user and their (possible) UserFineGroup.

**Only marking a fine as paid when a user has gone out of debt**
Take this hypothetical scenario. A user has a debt of 5 euros. Then, he gets (over time) fined for a total of 20 euros, bringing his total debt to 25 euros. His SudoSOS account then marks his balance as €-25,00 of which €20,00 are fines. If the user then deposits 20 euros into their account, his balance will now be €-5,00, of which €20,00 are fines. That does not make sense, but to keep the backend code simple, it was kept this way. If we wanted to remove this, we would have to fiddle with `Fine` objects or calculate/store the amount of fines in a different way. Of course, we can still simply "fix" this in the frontend by having the amount of fines be at most the users debt.

### CODE REVIEW

File: src/controller/debtor-controller.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
+/**
+ *  SudoSOS back-end API service.
+ *  Copyright (C) 2020  Study association GEWIS
```
Comment (CodeNamedRobin) [inline src/controller/debtor-controller.ts:3]: Denk dat deze wel geüpdatet moet worden?

File: src/controller/request/debtor-request.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @typedef HandoutFinesRequest
+ * @property {Array<integer>} userIds.required - Users to fine. If a user is not eligible for a fine, a fine of 0,00 will be handed out.
+ * @property {string} referenceDate - Reference date on to calculate balance for (and "now")
```
Comment (CodeNamedRobin) [inline src/controller/request/debtor-request.ts:22]: This sentence is not correct I believe. Also is the default "now"?

File: src/service/debtor-service.ts [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+   */
+  public static async deleteFine(id: number): Promise<void> {
+    const fine = await Fine.findOne({ where: { id }, relations: ['transfer', 'userFineGroup', 'userFineGroup.fines'] });
+    if (fine == null) return;
+
+    const { transfer, userFineGroup } = fine;
+
+    await Fine.remove(fine);
+    await Transfer.remove(transfer);
+    if (userFineGroup.fines.length === 1) await UserFineGroup.remove(userFineGroup);
```
Comment (CodeNamedRobin) [inline src/service/debtor-service.ts:270]: Maybe throw an error message here if length !== 0?

**Review by CodeNamedRobin — CHANGES_REQUESTED** [review]
Couple of small comments but other than that LGTM!

File: src/controller/debtor-controller.ts (hunk 1 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:3]: Ja, maar laten we dat doen in een andere PR. Dit is namelijk in de hele repository fout omdat het wordt ge-enforced door de linter.

File: src/service/debtor-service.ts (hunk 3 above)
Comment (Yoronex) [reply src/service/debtor-service.ts:270]: Niet eens, want een length === 0 kan niet (want je haalt de `UserFineGroup` op via `Fine`, dus de `UserFineGroup` heeft sowieso één fine) en bij een lengte van 2 is er nog minimaal één fine in de groep die je niet gaat verwijderen.

File: src/controller/request/debtor-request.ts (hunk 2 above)
Comment (Yoronex) [reply src/controller/request/debtor-request.ts]: Fixed in https://github.com/GEWIS/sudosos-backend/pull/42/commits/1261bbd43f4ec807b1443e718464cf93e1115b4a. Balances for fines are calculated on the reference date and now. If a user is in debt at both points in time, they are eligible for a fine.

**Review by CodeNamedRobin — APPROVED** [review]
Looks good to me!

File: src/controller/response/debtor-response.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * @property {Array<FineResponse.model>} fines - Fines that have been handed out
+ */
+export interface FineHandoutEventResponse extends BaseFineHandoutEventResponse {
+  fines: FineResponse[];
+}
+
+/**
+ * @typedef PaginatedFineHandoutEventResponse
+ * @property {PaginationResult.model} _pagination - Pagination metadata
+ * @property {Array<BaseFineHandoutEventResponse.model>} records - Returned fine handout events
```
Comment (JustSamuel) [inline src/controller/response/debtor-response.ts:64]: should be `Array.<BaseFineHandoutEventResponse>` iirc.

File: src/controller/response/debtor-response.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * @property {BaseUserResponse.model} createdBy - User that handed out the fines
+ */
+export interface BaseFineHandoutEventResponse extends BaseResponse {
+  referenceDate: string;
+  createdBy: BaseUserResponse;
+}
+
+/**
+ * @typedef {BaseFineHandoutEventResponse} FineHandoutEventResponse
+ * @property {Array<FineResponse.model>} fines - Fines that have been handed out
```
Comment (JustSamuel) [inline src/controller/response/debtor-response.ts:55]: should be `Array.<FineResponse>` iirc.

File: src/controller/balance-controller.ts [hunk 6]
```diff
@@ -95,6 +95,10 @@ export default class BalanceController extends BaseController {
    * @param {string} date.query - Timestamp to get balances for
    * @param {integer} minBalance.query - Minimum balance
    * @param {integer} maxBalance.query - Maximum balance
+   * @param {boolean} hasFine.query - Only users with(out) fines
+   * @param {integer} minFine.query - Minimum fine
+   * @param {integer} maxFine.query - Maximum fine
+   * @param {Array<string>} userTypes.query - User types (strings, not numbers)
```
Comment (JustSamuel) [inline src/controller/balance-controller.ts:101]: Mis handig om bij te zetten wat de opties zijn? dan zie je dat ook terug in de docs + frontend client.

File: src/controller/debtor-controller.ts [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        POST: {
+          policy: async (req) => this.roleManager.can(req.token.roles, 'get', 'all', 'Fine', ['*']),
+          handler: this.handoutFines.bind(this),
+        },
+      },
+    };
+  }
+
+  /**
+   * Get all fine handout events
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:75]: mist unieke @operationId

File: src/controller/debtor-controller.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+      res.json(await DebtorService.getFineHandoutEvents({ take, skip }));
+    } catch (error) {
+      this.logger.error('Could not return all fine handout event:', error);
+      res.status(500).json('Internal server error.');
+    }
+  }
+
+  /**
+   * Get all fine handout events
+   * @route GET /fines/{id}
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:109]: mist unieke @operationId

File: src/controller/debtor-controller.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+      res.json(await DebtorService.getSingleFineHandoutEvent(Number.parseInt(id, 10)));
+    } catch (error) {
+      this.logger.error('Could not return fine handout event:', error);
+      res.status(500).json('Internal server error.');
+    }
+  }
+
+  /**
+   * Delete a fine
+   * @route DELETE /fines/{id}
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:131]: mist unieke @operationId

File: src/controller/debtor-controller.ts [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    } catch (error) {
+      this.logger.error('Could not return fine handout event:', error);
+      res.status(500).json('Internal server error.');
+    }
+  }
+
+  /**
+   * Return all users that had at most -5 euros balance both now and on the reference date
+   * For all these users, also return their fine based on the reference date.
+   * @route GET /fines/eligible
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:162]: mist unieke @operationId

File: src/controller/debtor-controller.ts [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+      res.json(await DebtorService.calculateFinesOnDate(params));
+    } catch (error) {
+      this.logger.error('Could not calculate fines:', error);
+      res.status(500).json('Internal server error.');
+    }
+  }
+
+  /**
+   * Handout fines to all given users.
+   * @route POST /fines/handout
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:197]: mist unieke @operationId

File: src/controller/debtor-controller.ts [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    }
+  }
+
+  /**
+   * Delete a fine
+   * @route DELETE /fines/{id}
+   * @group debtors - Operations of the debtor controller
+   * @security JWT
+   * @param {integer} id.path.required - The id of the fine which should be deleted
+   * @returns {} 204 - Success
```
Comment (JustSamuel) [inline src/controller/debtor-controller.ts:135]: returns string

File: src/service/debtor-service.ts [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+    return DebtorService.asFineHandoutEventResponse(fineHandoutEvent);
+  }
+
+  /**
+   * Return all users that had at most -5 euros balance both now and on the reference date
+   * For all these users, also return their fine based on the reference date.
+   * @param userTypes List of all user types fines should be calculated for
+   * @param referenceDate Date to base fines on. If undefined, use now.
+   */
+  public static async calculateFinesOnDate({ userTypes, referenceDate }: CalculateFinesParams): Promise<UserToFineResponse[]> {
```
Comment (JustSamuel) [inline src/service/debtor-service.ts:140]: klopt deze jsdoc? is het niet een calculateFinesParams object als param?

File: src/service/debtor-service.ts [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+    }))[0];
+
+    const date = referenceDate || previousFineGroup?.createdAt || new Date();
+
+    const balances = await BalanceService.getBalances({
+      date,
+      ids: userIds,
+    });
+
+    const { fines: fines1, fineHandoutEvent: fineHandoutEvent1, emails: emails1 } = await getConnection().transaction(async (manager) => {
```
Comment (JustSamuel) [inline src/service/debtor-service.ts:190]: mis ergens duidelijk aangeven dat je een transaction maakt, want anders kan je vage transaction in transaction errors krijgen als iemand deze functie ook vanuit een transaction called.

**Review by JustSamuel — COMMENTED** [review]
I agree with the comments made in the timeline. However, it is quite a lot of code so its difficult to get a good overview / grasp of it. Made some small remarks here and there.

File: src/service/debtor-service.ts (hunk 13 above)
Comment (Yoronex) [reply src/service/debtor-service.ts:140]: Ja dit mag! Zie screenshot
image

File: src/controller/response/debtor-response.ts (hunk 4 above)
Comment (Yoronex) [reply src/controller/response/debtor-response.ts]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/response/debtor-response.ts (hunk 5 above)
Comment (Yoronex) [reply src/controller/response/debtor-response.ts]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/balance-controller.ts (hunk 6 above)
Comment (Yoronex) [reply src/controller/balance-controller.ts]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 7 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:75]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 8 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:110]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 9 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:133]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 10 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:165]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 11 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts:201]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/controller/debtor-controller.ts (hunk 12 above)
Comment (Yoronex) [reply src/controller/debtor-controller.ts]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

File: src/service/debtor-service.ts (hunk 14 above)
Comment (Yoronex) [reply src/service/debtor-service.ts:191]: Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6

**Pull request merged**